src/ServiceControl/MessageFailures/Api/EditFailedMessages.cs

```diff
@@ -0,0 +1,151 @@
+                    return HttpStatusCode.BadRequest;
+                }
+
+                if (string.IsNullOrWhiteSpace(edit.MessageBody) || edit.MessageHeaders == null)
+                {
+                    logging.WarnFormat("There is no message body on the edit-retry for MessageID {0}.", failedMessageId);
+                    return HttpStatusCode.BadRequest;
+                }
+
+                // Encode the body in base64 so that the new body doesn't have to be escaped
+                var base64String = Convert.ToBase64String(Encoding.UTF8.GetBytes(edit.MessageBody));
+                await Bus.SendLocal(new EditAndSend
+                {
+                    FailedMessageId = failedMessageId,
+                    NewBody = base64String,
+                    NewHeaders = edit.MessageHeaders.ToDictionary(kvp => kvp.Key, kvp => kvp.Value)
+                }).ConfigureAwait(false);
+
+                return HttpStatusCode.Accepted;
+            };
+        }
+
+        public LockedHeaderModificationValidator LockedHeaderModificationValidator { get; set; }
+
+        EditConfigurationModel GetEditConfiguration()
+        {
+            return new EditConfigurationModel
+            {
+                Enabled = Settings.AllowMessageEditing,
+                LockedHeaders = new[]
+                {
+                    Headers.MessageId,
+                    Headers.SagaId,
+                    Headers.CorrelationId,
+                    Headers.ControlMessageHeader,
+                    Headers.OriginatingSagaId,
+                    Headers.RelatedTo,
+                    Headers.ConversationId,
+                    Headers.MessageIntent,
+                    Headers.NServiceBusVersion,
+                    Headers.IsSagaTimeoutMessage,
+                    Headers.IsDeferredMessage,
+                    Headers.DelayedRetries,
+                    Headers.DelayedRetriesTimestamp,
+                    Headers.ImmediateRetries,
+                    Headers.ProcessingStarted,
+                    Headers.ProcessingEnded,
+                    "NServiceBus.ExceptionInfo.ExceptionType",
+                    "NServiceBus.ExceptionInfo.HelpLink",
+                    "NServiceBus.ExceptionInfo.Message",
+                    "NServiceBus.ExceptionInfo.Source",
+                    "NServiceBus.ExceptionInfo.StackTrace"
+                },
+                SensitiveHeaders = new[]
+                {
+                    Headers.RouteTo,
+                    Headers.DestinationSites,
+                    Headers.OriginatingSite,
+                    Headers.HttpTo,
+                    Headers.ReplyToAddress,
+                    Headers.ReturnMessageErrorCodeHeader,
+                    Headers.SagaType,
+                    Headers.OriginatingSagaType,
+                    Headers.TimeSent,
+                    "Header"
```

Comment:
Is there any explanation written down for why each of these of these headers should be editable?

Reply:
No there isn't. Where would we put such a thing? It's not the normal kind of doco we would write.

Reply:
It doesn't necessarily have to be doco, but I'd like to see an explanation and justification for each field. Previously we agreed that only custom headers would be editable and all system (NSB) headers would be locked. I'd like to understand why we are allowing editing for each of these fields.

Reply:
The following headers, currently categorized as `sensitive` contains user owned information, that's the reasoning behind the decision to allow them to be edited.

Gateway related headers might need to be edited in case a site is renamed

- Headers.DestinationSites
- Headers.OriginatingSite
- Headers.HttpTo

Routing headers might need to be renamed in case an endpoint is renamed

- Headers.RouteTo
- Headers.ReplyToAddress

Saga types related headers might be edited if a saga type changes over time

- Headers.SagaType
- Headers.OriginatingSagaType
- Headers.TimeSent

`Headers.ReturnMessageErrorCodeHeader` is used to store (if I understood correctly) the return value when using callbacks

@dbelcham in theory `"Header"` can be moved to locked headers, it was used as a prefix, thus is not a proper header. It's probably worth consulting a Core expert to better understand what this is.

Reply:
Thanks @mauroservienti, that helps a lot. Perhaps it would be good to group the list in this way, and include those explanations as comments?

I understand the justification for most of them, but I'm struggling a bit with the saga-related headers. Why would we need to change each of those?

Reply:
>include those explanations as comments?

@dbelcham what do you think?

>I understand the justification for most of them, but I'm struggling a bit with the saga-related headers. Why would we need to change each of those?

User saga type changed namespace for example due to a refactoring. I'm not sure how they are used by Core. I'm not how they are used, they are populated by Core but then as far as I can tell never used.

Reply:
> include those explanations as comments?

Is this something we normally do? I've never seen documentation about decisions included in code as a comment before in any of our codebases. My past experience says that this type of documentation will rot in the codebase over time so I would vote no to including it.

Reply:
I wonder if `Headers.SagaType` and `Headers.OriginatingSagaType` are used by SP/SI in some way? BTW do we know what the difference is between them?

I guess `Headers.TimeSent` is not related to sagas? Do we know how that's used and why it might need to be edited?

Reply:
If message processing depends on time and there is a large clock drift they may want to adjust that to accommodate clock drift

SagaType is the destination saga type of message I’d addressed to a saga. Originating Saga Type is if message originally come from a saga. I guess it’s used in the reply to originator scenario